TestDs3/Helpers/TestDs3ClientHelpers.cs

```diff
@@ -585,5 +586,404 @@ public void TestJobOnFailureEvent()
+            client
+                .Setup(c => c.BuildFactory(Stubs.Nodes))
+                .Returns(clientFactory.Object);
+            client
+                .Setup(c => c.PutBulkJobSpectraS3(MockHelpers.ItIsBulkPutRequest(Stubs.BucketName, ds3Objects, null)))
+                .Returns(new PutBulkJobSpectraS3Response(initialJobResponse));
+            client
+                .Setup(c => c.AllocateJobChunkSpectraS3(MockHelpers.ItIsAllocateRequest(Stubs.ChunkId1)))
+                .Returns(AllocateJobChunkSpectraS3Response.Success(Stubs.Chunk1(Stubs.NodeId2, false, false)));
+            client
+                .Setup(c => c.AllocateJobChunkSpectraS3(MockHelpers.ItIsAllocateRequest(Stubs.ChunkId2)))
+                .Returns(AllocateJobChunkSpectraS3Response.Success(Stubs.Chunk2(Stubs.NodeId2, false, false)));
+            client
+                .Setup(c => c.AllocateJobChunkSpectraS3(MockHelpers.ItIsAllocateRequest(Stubs.ChunkId3)))
+                .Returns(AllocateJobChunkSpectraS3Response.Success(Stubs.Chunk3(Stubs.NodeId1, false, false)));
+
+            const int timesToFail = 2;
+            var timesToFailLeft = 0;
+            node1Client
+                .Setup(c => c.PutObject(MockHelpers.ItIsPutObjectRequest(
+                    Stubs.BucketName, "hello", Stubs.JobId, 0L)))
+                .Callback(() =>
+                {
+                    if (timesToFailLeft < timesToFail)
+                    {
+                        timesToFailLeft++;
+                        throw new IOException();
+                    }
+                });
+
+            const int timesToFail2 = 2;
+            var timesToFailLeft2 = 0;
+            node2Client
+                .Setup(c => c.PutObject(MockHelpers.ItIsPutObjectRequest(
+                    Stubs.BucketName, "bar", Stubs.JobId, 15L)))
+                .Callback(() =>
+                {
+                    if (timesToFailLeft2 < timesToFail2)
+                    {
+                        timesToFailLeft2++;
+                        throw new IOException();
+                    }
+                });
+
+            var job = new Ds3ClientHelpers(client.Object).StartWriteJob(Stubs.BucketName, ds3Objects, null);
+            job.WithRetransmitFailingPutBlobs(2);
+
+            var dataTransfers = new ConcurrentQueue<long>();
+            var itemsCompleted = new ConcurrentQueue<string>();
+            job.DataTransferred += dataTransfers.Enqueue;
+            job.ItemCompleted += itemsCompleted.Enqueue;
+
+            job.Transfer(key => new MockStream(streams[key]));
+
+            node1Client.VerifyAll();
+            node2Client.VerifyAll();
+            clientFactory.VerifyAll();
+            client.VerifyAll();
+
+            CollectionAssert.AreEquivalent(new[] { 15L, 20L, 11L, 10L, 10L, 10L }, dataTransfers);
+            CollectionAssert.AreEquivalent(Stubs.ObjectNames, itemsCompleted);
+        }
+
+        [Test]
+        public void TestPutJobWithRetransmit2Blobs1Fail()
```

Comment:
A lot of the setup for the new tests are very similar (if not the same).  Can some of the common functionality be pulled out into a setup function?